Dynamic bearer token request and use

pbugni opened the pull request:
For aidbox, we need to make a round trip request to obtain a bearer token and then include it in the `Authorization` header on every subsequent request.

This PR wraps all `requests` calls in a specialized `requests.session` which obtains the token when necessary, before making any additional http requests to the FHIR server.

New configuration values required for use:
- FHIR_CLIENT_ID
- FHIR_CLIENT_SECRET

pbugni committed:
additional config vars needed for obtaining a live bearer token

pbugni committed:
abstraction to wrap requests.session with dynamically fetched bearer token

pbugni committed:
correct minor comment typo

pbugni committed:
refactor to use AuthSession to obtain a bearer token when necessary before making external requests to FHIR

pbugni committed:
must also use FHIR_SERVER_AUTH_URL given potential path differences for different vendors

pbugni committed:
remove unnecessary import

pbugni committed:
add logging for debugging

pbugni committed:
add don't verify to all FHIR calls given warnings

pbugni committed:
additional debugging logging

pbugni committed:
small typo; big result

pbugni committed:
don't raise on 988: InsecureRequestWarning

Commit:
more debugging logging

ivan-c commented on carl/modules/authsession.py:
```diff
@@ -0,0 +1,79 @@
+    # verify=False supress SSL warnings like curl --insecure  (NOT ADVISED)
+    current_app.logger.debug(f"token request to {FHIR_SERVER_AUTH_URL}")
+    current_app.logger.debug(f"with headers: {headers} and data: {body}")
+    response = requests.get(FHIR_SERVER_AUTH_URL, verify=False, headers=headers, data=body)
```
I think this needs to be POST
```suggestion
    response = requests.post(FHIR_SERVER_AUTH_URL, verify=False, headers=headers, data=body)
```

pbugni replied:
spot on!  overlooked the `curl --data` is an implicit POST.  thank you.

Commit:
must POST to /auth/token (duh!)

Commit:
must transmit as `json`

pbugni committed:
add required "content" attribute to CodeSystems (aidbox validation won't allow without)

pbugni committed:
add required "status" attribute to CodeSystems (aidbox validation won't allow without)

pbugni committed:
add required "content" and "status" attribute to CodeSystems (aidbox validation won't allow without)

pbugni committed:
clean up unnecessary debugging logging

pbugni committed:
add logging for next_page

pbugni committed:
handle relative path next page link

pbugni committed:
handle relative path next page link

pbugni committed:
comment cleanup

pbugni committed:
enable use on servers with and without FHIR_SERVER_AUTH_URL.
provide config where FHIR_SERVER doesn't have a valid certificat.

pbugni committed:
update github actions to use python-version 3.11

pbugni committed:
whitespace, as per black warning

pbugni committed:
pin pytest-cov version

pbugni committed:
whitespace only change.
new `black` complaint for all modules with comment at header - requires blank line to follow

pbugni committed:
remove coverage call and dependency from py.test due to bit rot and lack of use.

pbugni committed:
revert to python 3.9 given build/test errors on CI

pbugni committed:
revert to python 3.7 given build/test errors on CI

pbugni committed:
avoid test error by giving FHIR_SERVER_URL an empty string on default

pbugni committed:
lint check fails, can't locate a version older than 3.9.23

ivan-c commented on .github/workflows/lint.yaml:
```diff
@@ -15,7 +15,7 @@ jobs:
         uses: actions/setup-python@v1
         with:
-          python-version: 3.8
+          python-version: 3.9
```
I think we'll want to update the [python version in test.yaml](https://github.com/uwcirg/carl/blob/develop/.github/workflows/test.yaml#L9) too

That can wait, and I may have missed that you're keeping 3.7 until this is merged

ivan-c reviewed (approved):
looks good!!
